File: SecureBlackbox Samples/Public Key Crypto/python/public_key_crypto.py
# 
# SecureBlackbox 2022 Python Edition - Sample Project
# 
# This sample project demonstrates the usage of SecureBlackbox in a 
# simple, straightforward way. It is not intended to be a complete 
# application. Error handling and other checks are simplified for clarity.
# 
# www.nsoftware.com/secureblackbox
# 
# This code is subject to the terms and conditions specified in the 
# corresponding product license agreement which outlines the authorized 
# usage and restrictions.
# 

import sys
import string
from secureblackbox import *

input = sys.hexversion<0x03000000 and raw_input or input

def displayHelp():
    print(
        "NAME\n"
        "  publickeycrypto -- SecureBlackbox PublicKeyCrypto Demo Application\n\n"
        "SYNOPSIS\n"
        "  publickeycrypto -s/-v [-input input_file] [-output output_file] [-signature -signature_file]\n"
        "                  [-cert certificate_file] [-certpass certificate_password] [-encoding signature_encoding]\n\n"
        "DESCRIPTION\n"
        "  PublicKeyCrypto demonstrates the usage of PublicKeyCrypto from SecureBlackbox.\n"
        "  Used to sign and verify files.\n\n"
        "  The options are as follows:\n\n"
        "  -s            Sign input file and save to output \n\n" 
        "  -v            Verify signature using original file (input) \n\n" 
        "  -input        An input file to sign or verify (Required). \n\n"
        "  -output       Where the signed file will be saved (Required for signing). \n\n"
        "  -signature    An signature file to verify (Required on verifing). \n\n"
        "  -cert         The certificate used to sign files (Required).\n\n"
        "  -certpass     The password for the signing certificate (Required).\n\n"
        "  -encoding     The encoding of signature. Valid values:\n\n"
        "                  0 - DEFAULT\n"
        "                  1 - BINARY\n"
        "                  2 - BASE_64\n"
        "                  3 - COMPACT\n"
        "                  4 - JSON\n\n"
        "EXAMPLES\n"
        "  publickeycrypto -s -input C:\\cypto\\helloworld.txt -output C:\\cypto\\signature.dat -cert C:\\certs\\mycert.pfx -certpass mypassword \n\n"
        "  publickeycrypto -v -input C:\\cypto\\helloworld.txt -signature C:\\cypto\\signature.dat -cert C:\\certs\\mycert.pfx -certpass mypassword -encoding 2 \n\n"
    )

if (len(sys.argv) < 10):
    displayHelp()
    sys.exit(1)
else:
    crypto = PublicKeyCrypto()
    cm = CertificateManager()
    ckm = CryptoKeyManager()
    
    sign = False
    verify = False
    inputF = ""
    outputF = ""
    signatureF = ""
    certF = ""
    certpass = ""
    encoding = 0
    
    for x in range(len(sys.argv)):
        if (sys.argv[x].startswith("-")):
            if (sys.argv[x].lower() == "-s"):
                sign = True
            if (sys.argv[x].lower() == "-v"):
                verify = True
            if (sys.argv[x].lower() == "-input"):
                inputF = sys.argv[x+1]
            if (sys.argv[x].lower() == "-output"):
                outputF = sys.argv[x+1]
            if (sys.argv[x].lower() == "-signature"):
                signatureF = sys.argv[x+1]
            if (sys.argv[x].lower() == "-cert"):
                certF = sys.argv[x+1]
            if (sys.argv[x].lower() == "-certpass"):
                certpass = sys.argv[x+1]
            if (sys.argv[x].lower() == "-encoding"):
                encoding = int(sys.argv[x+1])

    if (not (sign or verify)):
        print("-s or -v is required.\n")
        displayHelp()
        sys.exit(1)

    if (sign and verify):
        print("-Use only one -s or -v parameter.\n")
        displayHelp()
        sys.exit(1)

    if (inputF == ""):
        print("-input is required.\n")
        displayHelp()
        sys.exit(1)
    
    if (sign and (outputF == "")):
        print("-output is required.\n")
        displayHelp()
        sys.exit(1)
    
    if (verify and (signatureF == "")):
        print("-signature is required.\n")
        displayHelp()
        sys.exit(1)
    
    if (certF == ""):
        print("-cert is required.\n")
        displayHelp()
        sys.exit(1)
    
    if (certpass == ""):
        print("-certpass is required.\n")
        displayHelp()
        sys.exit(1)
        
    try:
        cm.import_from_file(certF, certpass)
        ckm.set_cert_handle(cm.get_cert_handle())
        ckm.import_from_cert()
        crypto.set_key_handle(ckm.get_key_handle())
      
        if sign:
            crypto.set_output_encoding(encoding)
            crypto.sign_file(inputF, outputF, True)
            print("The file was signed successfully.");
        else:
            crypto.set_input_encoding(encoding)
            crypto.verify_detached_file(inputF, signatureF)

            if (crypto.get_signature_validation_result == svtSignerNotFound):
                print("Signer not found")
            elif (crypto.get_signature_validation_result == svtFailure):
                print("Signature verification failed")
            elif (crypto.get_signature_validation_result == svtCorrupted):
                print("Signature is invalid")
            elif (crypto.get_signature_validation_result == svtValid):
                print("Signature validated successfully.")
            else:
                print("Verification unknown.") 
    except Exception as e: 
        print(e)

File: SecureBlackbox Samples/Public Key Crypto/php/public_key_crypto.php
<?php
/*
 * SecureBlackbox 2024 PHP Edition - Sample Project
 *
 * This sample project demonstrates the usage of SecureBlackbox in a 
 * simple, straightforward way. It is not intended to be a complete 
 * application. Error handling and other checks are simplified for clarity.
 *
 * www.nsoftware.com/secureblackbox
 *
 * This code is subject to the terms and conditions specified in the 
 * corresponding product license agreement which outlines the authorized 
 * usage and restrictions.
 */
require_once('../include/secureblackbox_publickeycrypto.php');
require_once('../include/secureblackbox_cryptokeymanager.php');
require_once('../include/secureblackbox_certificatemanager.php');
require_once('../include/secureblackbox_const.php');
?>
<style>
table { width: 100% !important; }
td { white-space: nowrap; }
td input { width: 100%; }
td:last-child { width: 100%; }
</style>

<div width="90%">
  <form method=POST>
    <h2>Public key crypto Demo</h2>
    
    <h3>General Options</h3>
    <label><input type ="radio" checked="checked" name="comtype" value="Sign" /> Sign </label>
    <label><input type ="radio" name="comtype" value="Verify" /> Verify </label>
    <br/>
    <br/>

    <table>
      <tr><td>Input File:</td><td><input type=text name=inputFile value=""></td></tr>
      <tr><td>Output/Signature File:</td><td><input type=text name=outputFile value=""></td></tr>
    </table>
    <br/>
    <br/>

    <table>
      <tr>
        <td>Encoding:</td>
        <td>
          <select name="encoding">
            <option value="Binary">Binary</option>
            <option selected="selected" value="Base64">Base64</option>
            <option value="Compact">Compact</option>
            <option value="JSON">JSON</option>
          </select>
        </td>
      </tr>
    </table>
    <br/>
    <br/>

    <b>Signing Certificate</b>
    <table>
      <tr><td>Certificate File:</td><td><input type=text name=sCertFile value=""></td></tr>
      <tr><td>Password:</td><td><input type=password name=sCertPass value=""></td></tr>
    </table>
    <br/>
    <br/>

    <input type="submit" value="Sign/Verify" />
  </form>
</div><br/>

<?php
  if ($_SERVER['REQUEST_METHOD'] == "POST") {
    $crypto = new SecureBlackbox_PublicKeyCrypto();
    $certmgr = new SecureBlackBox_CertificateManager();
    $keymgr = new SecureBlackBox_CryptoKeyManager();
    
    try {
      // General options
      $inputFile = $_REQUEST['inputFile'];
      $outputFile = $_REQUEST['outputFile'];

      $comtype = $_POST["comtype"];
      if ($comtype == "Sign")
      {
        switch(trim($_REQUEST['encoding'])){
          case "Binary":  $crypto->setOutputEncoding(PUBLICKEYCRYPTO_OUTPUTENCODING_BINARY); break;
          case "Base64":  $crypto->setOutputEncoding(PUBLICKEYCRYPTO_OUTPUTENCODING_BASE_64); break;
          case "Compact":  $crypto->setOutputEncoding(PUBLICKEYCRYPTO_OUTPUTENCODING_COMPACT); break;
          case "JSON":  $crypto->setOutputEncoding(PUBLICKEYCRYPTO_OUTPUTENCODING_JSON); break;
          default: $crypto->setOutputEncoding(PUBLICKEYCRYPTO_OUTPUTENCODING_DEFAULT); break;
        }
      }
      else
      {
        switch(trim($_REQUEST['encoding'])){
          case "Binary":  $crypto->setInputEncoding(PUBLICKEYCRYPTO_INPUTENCODING_BINARY); break;
          case "Base64":  $crypto->setInputEncoding(PUBLICKEYCRYPTO_INPUTENCODING_BASE_64); break;
          case "Compact":  $crypto->setInputEncoding(PUBLICKEYCRYPTO_INPUTENCODING_COMPACT); break;
          case "JSON":  $crypto->setInputEncoding(PUBLICKEYCRYPTO_INPUTENCODING_JSON); break;
          default: $crypto->setInputEncoding(PUBLICKEYCRYPTO_INPUTENCODING_DEFAULT); break;
        }
      }

      // Signing options
      $certmgr->doImportFromFile($_REQUEST['sCertFile'], $_REQUEST['sCertPass']);
      $keymgr->setCertHandle($certmgr->getCertHandle());
      $keymgr->doImportFromCert();
      $crypto->setKeyHandle($keymgr->getKeyHandle());

      if ($comtype == "Sign")
      {
        $crypto->doSignFile($inputFile, $outputFile, TRUE);
        echo "<h2>The file was signed successfully</h2>";
      }
      else
      {
        $crypto->doVerifyDetachedFile($inputFile, $outputFile);

        switch($crypto->getSignatureValidationResult())
        {
          case PUBLICKEYCRYPTO_SIGNATUREVALIDATIONRESULT_VALID:  echo "<h2>Signature validated successfully</h2>";          break;
          case PUBLICKEYCRYPTO_SIGNATUREVALIDATIONRESULT_CORRUPTED:  echo "<h2>Signature is invalid</h2>";      break;
          case PUBLICKEYCRYPTO_SIGNATUREVALIDATIONRESULT_SIGNER_NOT_FOUND:  echo "<h2>Signer not found</h2>"; break;
          case PUBLICKEYCRYPTO_SIGNATUREVALIDATIONRESULT_FAILURE:  echo "<h2>Signature verification failed</h2>";        break;
          default: echo "<h2>Unknown</h2>";        break;
        }
      }
    }
    catch (exception $e) {
      echo "<h2>Signing Failure (Details Below)</h2><p>" . $e->getMessage() . "</p>";
    }
  }
?>